R-notebooks/covidhub_evaluation.Rmd

```diff
@@ -49,57 +49,97 @@ forecasters = c("CMU-TimeSeries",
 to using the Zoltar API, and soon it should have an "as of" parameter, so then 
 we shouldn't need to do this. 
 
-```{r, eval = FALSE}
-forecast_dates_comparable = vector("list", length = length(forecasters)) 
+```{r}
+
 forecast_dates_cmu = forecast_dates[[1]]
+new_dates = list()
 for (i in 1:length(forecasters)) {
   given_dates = forecast_dates[[i]]
-  for (j in 1:length(forecast_dates_cmu)) {
-    # Find the last forecast date before the current CMU one
-    given_date = given_dates[max(which(given_dates <= forecast_dates_cmu[j]))]
-    
-    # If the dates match exactly, or the given date falls on a Sunday and the
-    # CMU date falls on a Monday of the same epiweek, the call it comparable
-    if (!is.na(given_date) &&
-        (given_date == forecast_dates_cmu[j] ||
-        (as.Date(forecast_dates_cmu[j]) - as.Date(given_date) == 1 &&
-         wday(forecast_dates_cmu[j]) == 2 && wday(given_date) == 1))) {
-      forecast_dates_comparable[[i]][j] = given_date
-    }
-    
-    # Otherwise call it NA
-    else forecast_dates_comparable[[i]][j] = NA
+  # If the dates match exactly, or the given date falls on a Sunday and the
+  # CMU date falls on a Monday of the same epiweek, then call it comparable...
+  dates_comp = given_dates[(given_dates %in% forecast_dates_cmu | 
+                            ((given_dates + 1) %in% forecast_dates_cmu) &
+                            wday(given_dates) == 1)]
+  
+  # ...but if there is an exact match on dates, ignore predictions made on the
+  # previous day
+  extra_dates = dates_comp[(dates_comp + 1) %in% dates_comp]
+  dates_comp = as_date(setdiff(dates_comp, extra_dates))
+  if(forecasters[[i]] %in% names(seen_dates)){
+    forecaster_idx = which(names(seen_dates) == forecasters[[i]])
+    dates_comp = dates_comp[!(dates_comp %in% seen_dates[[forecaster_idx]])]
   }
+  new_dates[[i]] = dates_comp
 }
+names(new_dates) = forecasters
 ```
 
 Now get predictions for each forecaster, looping over forecast dates from the
 CMU-TimeSeries model. **TODO**: this part is very very slow. Maybe (hopefully) 
 by changing to use the Zoltar API, this will be much faster.
 
-```{r, eval = FALSE}
-predictions_cards = vector("list", length = length(forecasters))
+```{r}
+predictions_cards_new = vector("list", length = length(forecasters))
 for (i in 1:length(forecasters)) {
   cat(forecasters[i], "...\n")
-  predictions_cards[[i]] = tryCatch({
+  predictions_cards_new[[i]] = tryCatch({
     get_covidhub_predictions(forecasters[i], 
-                             na.omit(forecast_dates_comparable[[i]]), 
+                             new_dates[[i]], 
                              geo_type = "state")
   },
   error = function(e) cat(e$message))
 }
+old_forecaster_idxs = match(forecasters, names(seen_dates))
+
+# Combine old and new predictions cards
+predictions_cards_comb = vector("list", length = length(forecasters))
+for (i in 1:length(forecasters)){
+  if (is.na(old_forecaster_idxs[i])){
```

Comment:
Do you want `i %in% old_forecaster_idxs` instead?  My understanding is that `old_forecaster_idxs` is a list of index positions such as `c(1, 3, 7, 9)` corresponding to forecasters already seen.  So `old_forecaster_idxs[i]` asks for the `i`th member of the list rather than whether `i` is in the list.

Reply:
The result from `match` is the same length as `forecasters` and specifies which value in the 2nd vector matches the value in the 1st vector (with NA representing no match).  

For example, `match(c('foo', 'bar', 'baz'), c('baz', 'foo'))` would be `c(2, NA, 1)`.

so, `old_forecaster_idxs[i]` being `NA` means there is no match.